lib/db/observed.go

```diff
@@ -0,0 +1,265 @@
+				return nil, err
+			}
+			continue
+		}
+		// Here we expect the length to match, coming from the device index.
+		deviceID := protocol.DeviceIDFromBytes(keyDev)
+		if err := db.collectPendingFolder(iter.Key(), deviceID, res); err != nil {
+			return nil, err
+		}
+	}
+	return res, nil
+}
+
+// PendingFoldersForDevice drops any invalid entries from the database as a side-effect.
+func (db *Lowlevel) PendingFoldersForDevice(device protocol.DeviceID) (map[string]map[protocol.DeviceID]ObservedFolder, error) {
+	prefixKey, err := db.keyer.GeneratePendingFolderKey(nil, device[:], nil)
+	if err != nil {
+		return nil, err
+	}
+	iter, err := db.NewPrefixIterator(prefixKey)
+	if err != nil {
+		return nil, err
+	}
+	defer iter.Release()
+	res := make(map[string]map[protocol.DeviceID]ObservedFolder)
+	for iter.Next() {
+		if err := db.collectPendingFolder(iter.Key(), device, res); err != nil {
+			return nil, err
+		}
+	}
+	return res, nil
+}
+
+func (db *Lowlevel) collectPendingFolder(key []byte, device protocol.DeviceID, res map[string]map[protocol.DeviceID]ObservedFolder) error {
+	bs, err := db.Get(key)
+	if err != nil {
+		return err
+	}
+	var of ObservedFolder
+	folderID := string(db.keyer.FolderFromPendingFolderKey(key))
+	if len(folderID) < 1 {
+		return db.deleteInvalidPendingFolder(key)
+	}
+	if err := of.Unmarshal(bs); err != nil {
+		return db.deleteInvalidPendingFolder(key)
+	}
+	if _, ok := res[folderID]; !ok {
+		res[folderID] = make(map[protocol.DeviceID]ObservedFolder)
+	}
+	res[folderID][device] = of
+	return nil
+}
+
+// deleteInvalidPendingFolder logs a warning message before dropping the given entry.
+// That's the only possible "repair" measure and appropriate for the importance of pending
+// entries.  They will come back soon if still relevant.
+func (db *Lowlevel) deleteInvalidPendingFolder(key []byte) error {
+	l.Warnf("Invalid pending folder entry, deleting from database: %x", key)
+	err := db.Delete(key)
+	return err
+}
+
+// Set of devices for which pending folders are allowed, but not specific folder IDs
+type DropListObserved map[protocol.DeviceID]map[string]struct{}
+
+func NewDropListObserved() DropListObserved {
+	return make(DropListObserved)
+}
+
+func (dl DropListObserved) MarkDevice(device protocol.DeviceID) map[string]struct{} {
+	folders, ok := dl[device]
+	if !ok {
+		dl[device] = nil
+	}
+	return folders
+}
+
+func (dl DropListObserved) UnmarkDevice(device protocol.DeviceID) {
+	delete(dl, device)
+}
+
+func (dl DropListObserved) MarkFolder(folder string, devices []protocol.DeviceID) {
+	for _, dev := range devices {
+		folders := dl.MarkDevice(dev)
+		if folders == nil {
+			folders = make(map[string]struct{})
+			dl[dev] = folders
+		}
+		folders[folder] = struct{}{}
+	}
+}
+
+// shouldDropPendingDevice defines how the drop-list is interpreted for pending devices
+func (dl DropListObserved) shouldDropPendingDevice(key []byte, keyer keyer) bool {
```

Comment:
This should really be `shouldDropPendingDevice(device protocol.DeviceID)` and outsource keying magic to the caller, that's not the responsibility of the drop list thingy to do...

Reply:
The main reason for moving it to a utility function was to improve readability for the case where the key is unintelligible and no proper IDs can be extracted to compare against the drop list.  These two `shouldDrop...` functions essentially define the semantics of that type.  I agree that dropping invalid entries is not exactly part of that, but it was factored out of the `CleanPending...` functions as a whole to simplify the loop there.